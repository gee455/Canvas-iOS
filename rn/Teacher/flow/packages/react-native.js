//
// Copyright (C) 2016-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

// @flow

declare var console: typeof console & { disableYellowBox: boolean }

export type PanResponderGestureState = {
  stateID: number,
  moveX: number,
  moveY: number,
  x0: number,
  y0: number,
  dx: number,
  dy: number,
  vx: number,
  vy: number,
}

export type KeyboardEventData = {
  endCoordinates: {
    width: number,
    height: number,
    screenX: number,
    screenY: number,
  },
}
