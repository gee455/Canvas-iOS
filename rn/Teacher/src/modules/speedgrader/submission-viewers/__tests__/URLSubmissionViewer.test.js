//
// Copyright (C) 2016-present Instructure, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

// @flow

import { Image, Linking } from 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'
import URLSubmissionViewer from '../URLSubmissionViewer'
import setProps from '../../../../../test/helpers/setProps'
import explore from '../../../../../test/helpers/explore'

const t = {
  ...require('../../../../__templates__/submissions.js'),
  ...require('../../../../__templates__/attachment.js'),
}

Image.getSize = jest.fn()
const mockOpenURL = jest.fn()
jest.mock('Linking', () => ({ openURL: mockOpenURL }))
jest.mock('TouchableOpacity', () => 'TouchableOpacity')

const submission = t.submission({
  attachment: t.attachment(),
  submission_type: 'online_url',
  url: 'https://my-blog-about-rockets.com/2017/6/1/rockets-are-splosive',
})

test('url submission renders correctly', () => {
  const tree = renderer.create(
    <URLSubmissionViewer submission={submission} drawerInset={0} />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

test('on new url received, fetch image height if attachment', () => {
  const rendered = renderer.create(
    <URLSubmissionViewer submission={submission} drawerInset={0} />
  )
  const fetch = jest.fn()
  rendered.getInstance().fetchImageSize = fetch

  setProps(rendered, { submission: { ...submission, attachments: [] } })
  expect(fetch).not.toHaveBeenCalled()

  const newPost = 'https://my-blog-about-rockets.com/2017/6/2/liftoff'
  const oldAttachments = submission.attachments || []
  const updated = {
    ...submission,
    attachments: [{ ...oldAttachments[0], url: newPost }],
  }
  setProps(rendered, { submission: updated })
  expect(fetch).toHaveBeenCalledWith(newPost)
})

test('on scrollview layout, the state is updated', () => {
  const rendered = renderer.create(
    <URLSubmissionViewer submission={submission} drawerInset={0} />
  )

  rendered.getInstance().onScrollViewLayout(
    { nativeEvent: { layout: { width: 200, height: 200 } } }
  )

  expect(rendered.getInstance().state.size).toEqual({ width: 200, height: 200 })
})

test('loaded image sets aspect ration for rendered image view', () => {
  const rendered = renderer.create(
    <URLSubmissionViewer submission={submission} drawerInset={0} />
  )

  rendered.getInstance().imageSizeLoaded(800, 600)

  expect(rendered.getInstance().state.aspectRatio).toEqual(800 / 600)
})

test('changing the fetches the dimensions', () => {
  const url = 'https://my-blog-about-rockets.com/2017/6/2/propulsion_ftw'
  const attachments = submission.attachments || [t.attachment()]
  const newSub = {
    ...submission,
    attachments: [{ ...attachments[0], url }],
  }

  const rendered = renderer.create(
    <URLSubmissionViewer submission={submission} drawerInset={0} />
  )

  setProps(rendered, { submission: newSub })
  expect(Image.getSize).toHaveBeenCalledWith(url, rendered.getInstance().imageSizeLoaded)
})

test('tapping the url opens the link in Safari', () => {
  const tree = renderer.create(
    <URLSubmissionViewer submission={submission} drawerInset={0} />
  ).toJSON()

  const urlButton = explore(tree).selectByID('url-submission-viewer.url')
  urlButton && urlButton.props.onPress()

  expect(Linking.openURL).toHaveBeenCalledWith(submission.url)
})
