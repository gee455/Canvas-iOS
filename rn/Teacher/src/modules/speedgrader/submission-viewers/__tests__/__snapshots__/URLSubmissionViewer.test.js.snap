// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`url submission renders correctly 1`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "flex": 1,
    }
  }
  contentInset={
    Object {
      "bottom": 0,
    }
  }
>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      testID="url-submission-viewer.explanation"
    >
      This submission is a URL to an external page. We've included a snapshot of what the page looked like when it was submitted.
    </Text>
    <TouchableOpacity
      accessibilityTraits="button"
      hitSlop={
        Object {
          "bottom": 15,
          "left": 15,
          "right": 15,
          "top": 15,
        }
      }
      onPress={[Function]}
      style={
        Object {
          "alignSelf": "center",
          "paddingVertical": 12,
        }
      }
      testID="url-submission-viewer.url"
    >
      <View>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "fontFamily": ".SFUIDisplay",
              },
              Object {
                "color": "#2D3B45",
                "fontSize": 16,
              },
              Array [
                Object {
                  "color": "#374A59",
                  "fontSize": 14,
                },
                Object {
                  "fontFamily": ".SFUIDisplay-bold",
                },
                Object {
                  "color": "#374A59",
                },
                Object {},
                Object {
                  "fontSize": 18,
                },
              ],
              Object {
                "fontFamily": ".SFUIDisplay",
              },
            ]
          }
        >
          https://my-blog-about-rockets.com/2017/6/1/rockets-are-splosive
        </Text>
      </View>
    </TouchableOpacity>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "fontFamily": ".SFUIDisplay",
          },
          Object {
            "color": "#7F91A7",
            "fontSize": 16,
            "lineHeight": 23,
          },
          Object {
            "alignSelf": "center",
          },
        ]
      }
    >
      Preview Unavailable
    </Text>
  </View>
</RCTScrollView>
`;
